invenio_drafts_resources/services/draft.py

```diff
@@ -71,85 +62,141 @@ def draft_cls(self):
+        self._patch_data(record, validated_data)
+        draft = self.draft_cls.create(record.dumps(), id_=record.id,
                                       fork_version_id=record.revision_id)
         db.session.commit()  # Persist DB
         if self.indexer:
             self.indexer.index(draft)
 
-        return self.config.resource_unit_cls(pid=pid, record=draft)
+        return self.resource_unit(pid=pid, record=draft, links=None)
+
+    def _publish_revision(self, validated_data, pid):
+        """Publish draft of existing record (edition)."""
+        pid, record = self.pid_manager.resolve(pid.pid_value)
+        record.clear()
+        record.update(validated_data)
+        record.commit()
+
+        return record
+
+    def _publish_new_version(self, validated_data, draft):
+        """Publish draft of a new record."""
+        record = self.record_cls.create(validated_data, id_=draft.id)
+        conceptrecid = record['conceptrecid']
+
+        # FIXME: This implies two db queries for the same in pid_manager
+        if self.pid_manager.is_first_version(
+            pid_value=conceptrecid
+        ):
+            self.pid_manager.register_pid(record['conceptrecid'])
+
+        self.pid_manager.register_pid(record['recid'])
+
+        return record
 
     def publish(self, id_, identity):
         """Publish a draft."""
         self.require_permission(identity, "publish")
         # Get draft
-        pid, draft = self.resolve_draft(id_=id_)
-        # Validate and create record, register PID
-        data = draft.dumps()
+        pid, draft = self.pid_manager.resolve(id_, draft=True)
         # Validate against record schema
-        validated_data = self.data_validator.validate(data)
-        # Use same UUID
-        record = self.record_cls.create(validated_data, id_=draft.id)
-        pid.register()
+        validated_data = self.data_validator.validate(draft.dumps())
+        # Publish it
+        if self.pid_manager.is_published(pid):  # Then we publish a revision
+            record = self._publish_revision(validated_data, pid)
+        else:
+            record = self._publish_new_version(validated_data, draft)
         # Remove draft
-        draft.delete()
+        draft.delete(force=True)
         db.session.commit()  # Persist DB
         # Index the record
         if self.indexer:
             self.indexer.delete(draft)
             self.indexer.index(record)
 
-        return self.resource_unit(pid=pid, record=record)
+        return self.resource_unit(pid=pid, record=record, links=None)
+
+    def new_version(self, id_, identity):
+        """Create a new version of a record."""
+        self.require_permission(identity, "create")
+        # Get record
+        pid, record = self.pid_manager.resolve(id_)
+        # Validate and create a draft, register PID
+        data = record.dumps()
+        # Validate against record schema
+        validated_data = self.data_validator.validate(data, partial=True)
```

Comment:
Isn't it here better to have strict validation as we start from a record?

Reply:
I actually wonder if we should validate it? It's a nice-to-do but the record was already validated when publishing... WDYT?

Reply:
It makes sense not to validate :)